### PR TITLE
[core] Update properties framework

### LINKED ISSUE #1204: [core] Allow numeric properties in XML to be within an unbounded range
Report thresholds for metrics for example usually only have the constraint that they must be positive. It makes no sense to artificially restrict that, and it obliges us to take an arbitrary decision ("is 2000 enough? should I add a zero?"). We could add some methods `mustBeAbove` and `mustBeBelow` to the property builders and implement some logic in the NumericPropertyModule to have unbounded intervals.

* When defining a property programmatically, one can use `NumericConstraint.positive()`
* But when defining a property in XML, you always have to provide `min` and `max`:
```xml
<property name="intProp" type="Integer" value="3" min="1" max="20" description="An IntegerProperty."/>
```
  * this will always build a range constraint

### LINKED ISSUE #1432: [core] Implementation plan for property framework changes
This issue serves as easy reference for PRs, and as a discussion for the changes described [on the wiki](https://github.com/pmd/pmd/wiki/Property-framework-7-0-0). Replaces #1415 

## Implementation plan

<details>
  <summary>Due on 6.10.0 :heavy_check_mark: done</summary>
  <p>

- [x] **(#1420)** Deprecate all classes and methods to be removed
- [x] **(#1470)** Introduce a PropertyFactory, a PropertyConstraint, etc to get a property creation API
  that is as capable as the current framework, and source compatible with the 7.0.0 API. 
  * This doesn't need to implement anything about PropertyXmlSyntax, and doesn't entail changes to the PropertyDescriptor interface. It could be integrated to the current property framework seamlessly
- [x] Deprecate all concrete property types, migrate internal usages to use PropertyFactory
  (should allow us to spot some bugs, enrich API, etc):
    - [x] #1420
      - [x] MethodMultiProperty
      - [x] MethodProperty
      - [x] TypeMultiProperty
      - [x] TypeProperty
      - [x] FileProperty
    - [x] #1485 IntegerProperty
    - [x] #1486 
      - [x] DoubleProperty
      - [x] DoubleMultiProperty
      - [x] FloatMultiProperty
      - [x] FloatProperty
      - [x] IntegerMultiProperty
    - [x] #1487 RegexProperty
    - [x] #1501
      - [x] StringMultiProperty
      - [x] StringProperty
    - [x] #1502 
      - [x] BooleanMultiProperty
      - [x] BooleanProperty
    - [x] #1505 
      - [x] EnumeratedMultiProperty
      - [x] EnumeratedProperty
    - [x] #1506  
      - [x] CharacterMultiProperty
      - [x] CharacterProperty
      - [x] LongMultiProperty
      - [x] LongProperty

</p>
</details>

Due on 7.0.0

- [x] **Removing StatisticalRule** is a prerequisite, because its properties can not be converted to the framework
  - [x] #1750
  - [x] #1772
  - [x] #2005
- [x] #2450

### Probably for PMD 7
* #4301
* #4303

### Postponed
* #4302
* #4304
* #4305
* #4306
* #3869 
* #1204

### LINKED ISSUE #4301: [core] Remove deprecated property concrete classes
* And migrate last old property usages 
* This is around the classes in package `net.sourceforge.pmd.properties`
* :information_source: Note: There is a branch by @oowekyala that might contain some of the tasks above: https://github.com/pmd/pmd/compare/pmd/7.0.x...oowekyala:pmd:1432-property-framework-upgrade-pmd7

- [ ] Remove deprecated methods around (Abstract)PropertySource (see #1322)
  * originally from #881

### LINKED ISSUE #4302: [core] Migrate Property Framework API to Java 8
* "The constraint mechanism will be completely unlocked with 7.0.0, since we’ll be migrating our API to Java 8."
* :information_source: Note: There is a branch by @oowekyala that might contain some of the tasks above: https://github.com/pmd/pmd/compare/pmd/7.0.x...oowekyala:pmd:1432-property-framework-upgrade-pmd7

### LINKED ISSUE #4303: [doc] Document new property framework
* Document all these cool things that are already in pmd/7.0.x
* Review/Update existing documentation
  * from pmd6: https://pmd.github.io/latest/pmd_userdocs_extending_defining_properties.html
  * from pmd7: http://docs.pmd-code.org/pmd-doc-7.0.0-SNAPSHOT/pmd_userdocs_extending_defining_properties.html
* See wiki: https://github.com/pmd/pmd/wiki/Property-framework-7-0-0
* Update documentation generator to take PropertyConstraint into account
* :information_source: Note: There is a branch by @oowekyala that might contain some of the tasks above: https://github.com/pmd/pmd/compare/pmd/7.0.x...oowekyala:pmd:1432-property-framework-upgrade-pmd7

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

Update the property framework for PMD 7.
- remove deprecated API
- turn PropertyDescriptor into a class because it only needs a single implementation now
- prepare the framework to support new XML syntax in the future. This PR does not change the XML schema, but changes its interpretation in the following places:
   - the `delimiter` attribute for property definitions is deprecated, produces a warning and its value is ignored.
   - the `min` and `max` attributes for property definitions of numeric properties are no longer required, and can now appear separately, or not at all.

Maybe the most arbitrary change is that I removed the ability to configure the delimiter for multi-value properties, and changed the default delimiter to a comma (`,`) for all properties. The default delimiter was previously inconsistent as string properties used `|` for some reason. Many properties were overwriting this with a comma, in an inconsistent and unpredictable manner. There is no change in expressivity as the comma may now be escaped by a backslash. The actual reason I did this is because having a custom delimiter complicates the round-tripping through RulesetWriter a lot. Also I expect that the new syntax for list properties (which is not in this PR) will anyway become a nicer way to write list properties when writing a comma-delimited list is ugly.

The API for rules (PropertyFactory) is unchanged except for a couple of new convenience methods, to create enumerated properties from an Enum class.
'

## Related issues

- Fixes #1432 
- Fixes #1204 
- fix #4302 
- fix #4301
- fix #4303
- relates to #4305: this does not implement any new syntax, but sets up infrastructure to add this later without breaking everything. We should revert [84cba78](https://github.com/pmd/pmd/pull/4380/commits/84cba78e645ac0c586f6a64c532efefeb0d1be7a).

## Ready?

- [x] Review that not too many unit tests are missing
- [x] Review doc pages -> API didn't really change so probably little to do. Ref #4303
- [x] Add tests for #1204
- [x] Release notes
- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to pmd/7.0.x:<br />
This changeset changes 6 violations,<br />
introduces 1 new violations, 0 new errors and 0 new configuration errors,<br />
removes 1 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/142c975a3c99482a858b0f137fb75aa3/diff1/index.html">Full report</a></td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to master:<br />
This changeset changes 50624 violations,<br />
introduces 34682 new violations, 12 new errors and 0 new configuration errors,<br />
removes 142045 violations, 4 errors and 7 configuration errors.<br />
<a href="https://chunk.io/pmd/142c975a3c99482a858b0f137fb75aa3/diff2/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-core/src/main/java/net/sourceforge/pmd/RuleSetWriter.java [hunk 1]
```diff
@@ -255,95 +257,64 @@ private Element createRuleSetReferenceElement(RuleSetReference ruleSetReference)
...
+
+    private <T> Element createPropertyDefinitionElementBR(PropertyDescriptor<T> propertyDescriptor, @NonNull PropertyTypeId typeId) {
 
-        return propertyElement;
+        final Element element = createPropertyValueElement(propertyDescriptor, propertyDescriptor.defaultValue());
+
+        SchemaConstants.NAME.setOn(element, propertyDescriptor.name());
+        SchemaConstants.PROPERTY_TYPE.setOn(element, typeId.getStringId());
+        SchemaConstants.DESCRIPTION.setOn(element, propertyDescriptor.description());
+        // TODO support property constraints in XML
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/RuleSetWriter.java:316]: TODO support property constraints in XML

File: pmd-core/src/main/java/net/sourceforge/pmd/properties/PropertyDescriptor.java [hunk 2]
```diff
@@ -4,193 +4,179 @@
...
+    }
 
 
     /**
-     * Formats the object onto a string suitable for storage within the property map.
-     *
-     * @param value Object
-     *
-     * @return String
+     * TODO port tests to use the mapper directly.
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/properties/PropertyDescriptor.java:125]: TODO port tests to use the mapper directly.

**Review by adangel — APPROVED** [review]
Thanks! I finally got around to go through the changes.

I'm planning to work on the TODOs/comments myself in the next couple of days and merge it afterwards.

**Comment by oowekyala** [comment]
Thanks for finishing this @adangel !

**Pull request merged**